### DISCUSSION TITLE
Cargo containers for fighter bays.

### DISCUSSION THREAD

**Discussion by ghost** [endless-sky/endless-sky#9891] [Ideas]
If someone doesn't use the fighter bays for actual fighting, currently there's a Boxwing "fighter", which they can place in there and use them to have extra cargo space.
But, being a ship, they still each require a pilot (who has to be paid), and are listed in the list of ships owned by player (which can get quite cluttered, when there's a lot of them).

So, I think it could be a good idea to be able to buy a straightforward container for extra cargo, which would take a place of a fighter craft, but not be a fighter itself (so it wouldn't require a pilot, and wouldn't be listed as a ship).
Sort of like a bigger Cargo Expansion, just placed in the Fighter Bay, and providing 80 tons of cargo space (since that's how much the Boxwing has). Maybe a little more, because it wouldn't need to have engines and other things a ship does.

Yeah, if someone doesn't use fighters, they likely wouldn't use a dedicated carrier ship (like Skein or Navy Carrier), but there are also non-carrier ships having fighter bays, like Mule, Bactrian and Kestrel - those can be used for other purposes, and still have their fighter bays used for cargo.

**Comment by Saugia** [comment]
This is kind of in-line with the thoughts of having Fighter Bays as purchasable outfits for hardpoints. Generally, we want a ship's hardpoints to be static. If a ship has a fighter bay, it should be used for fighters. If a ship has a gun mount, it is for guns. Turret, for turrets, etc.. We don't really want to give hardpoints multiple dedicated uses.

The Boxwing case is certainly interesting and having fighters use fighter bays for cargo is actually one style of playing the game, yes. But we don't want to overlap hardpoint purposes. If a ship has dedicated fighter bays and you want to use it for cargo, but want more cargo without the fighter restrictions, you should look at a larger freight ship for cargo space that does not have fighter bays. We don't want to "optimize" carrier ships for cargo haulage, since that isn't their intended usage, and could get messy with balance.

The restrictions fighters bring as cargo pods is intended - we don't want to have a workaround that makes that space purely efficient in something that isn't fighter usage.

**Comment by ghost** [comment]
> We don't want to "optimize" carrier ships for cargo haulage, since that isn't their intended usage, and could get messy with balance.

I get where you're coming from, like i said, I was thinking not about the dedicated carrier-type ships, but a ship that people might want to buy for different reasons.

F.ex. I'm now working on getting the Bactrian license, planning to buy one and use it as my flagship for capturing purposes (all those bunks) - I'd still buy it if the Fighter bays weren't there, but since they are, I probably will use them for cargo (as I think fighters are too fragile and I'd probably have to micromanage and/or replace them too much to make it worth it)

Or Kestrel, I bought one (non-carrier version) to serve as a regular heavy warship, but (same reasoning as for Bactrian).

Regardless, I guess making those hypothetical cargo containers work only with let's call them hybrid ships could be too much of a hassle, programming-wise (even if you agreed with my reasoning, which you don't), so that's fair.

**Reply by Saugia** [reply]
I will say, although it may not be something looking at vanilla, and since it is more code-based, something you could maybe look into is a plug-in that has variants of carrier ships that, instead of fighter bays, have cargo space in place of them. Or make one if that doesn't exist already. If you're a plug-in person. Just because these ideas may not fit best with vanilla practice doesn't mean they wouldn't be welcome outside of it, especially as add-ons.

**Reply by Hecter94** [reply]
> F.ex. I'm now working on getting the Bactrian license, planning to buy one and use it as my flagship for capturing purposes (all those bunks) - I'd still buy it if the Fighter bays weren't there, but since they are, I probably will use them for cargo (as I think fighters are too fragile and I'd probably have to micromanage and/or replace them too much to make it worth it)

You could always set up those fighters for missile defense or set them up as missile boats if you want some more offensive support. 
On a Bactrian, for example, being able to deploy a couple of extra missile defense turrets on a button press can be pretty effective, and in any of those two roles, they don't tend to die much, if at all.

**Reply by ghost** [reply]
I had a (very) brief look at how plugins and outfits are made. It doesn't seem possible to make an Outfit going into Fighter Bay slot (so nothing like what I suggested in my first post could be easily done), so I guess if I tried it, my best bet would be making, via a plug-in, the Bay-less versions of those ships (like you suggested Saugia) , or a version of Boxwing that doesn't need a pilot, has no engine, etc.
(technically still a fighter, so I guess the player would have to be careful about not launching it by accident, as it would have no way to return)

If I tried doing either of those, would it be fine if I reused the existing graphic assets for that? (as I have zero drawing skills)

---

Or try the defensive use of fighters as suggested by Hecter94, I guess.

**Reply by Saugia** [reply]
You can certainly re-use vanilla assets as you wish, so long as you properly credit them in a copyright/credits file. Take a look at how other plug-ins do it over at our [plug-ins page](https://endless-sky.github.io/plugins.html), or alternatively look at how derivative assets are attributed in vanilla's copyright file.

But yeah, plug-ins are data file focused. If you wanted to change code, you'd have to do it in a separate repository.

**Reply by lumbar527** [reply]
> I had a (very) brief look at how plugins and outfits are made. It doesn't seem possible to make an Outfit going into Fighter Bay slot (so nothing like what I suggested in my first post could be easily done), so I guess if I tried it, my best bet would be making, via a plug-in, the Bay-less versions of those ships (like you suggested Saugia) , or a version of Boxwing that doesn't need a pilot, has no engine, etc.

What about autonomous fighters? If they had no sprite then you couldn't destroy them though you would still be able to deploy and forget them accidentally.

**Reply by ghost** [reply]
> What about autonomous fighters? If they had no sprite then you couldn't destroy them though you would still be able to deploy and forget them accidentally.

Yeah, I noticed something about making pilotless ships, when checking how they are made, would probably go in that direction.
(since my main problem was paying the pilot for doing nothing, and the guy would be bored out of their mind, if they were real ;) that would solve it)

Why would the player want to destroy them, however?

**Reply by lumbar527** [reply]
> Why would the player want to destroy them, however?

It would be if they accidentally hit `d` and forgot to hit it again before they jumped.

**Reply by ghost** [reply]
> It would be if they accidentally hit `d` and forgot to hit it again before they jumped.

That's a good point. If I do it, I may leave the engine in, so that it may return if needed. Maybe make a new outfit with a weaker engine, or just increase the mass (it would be slower either way), if trying to balance it a little.

Who is, in universe, the creator of Boxwing, btw? I'm thinking who should sell the new version. Since they are sold in the south, my guess would be Tarazed.
So, I'm thinking to either add it to Tarazed and Free Worlds shipyards, or just Free Worlds, in the latter case presenting it in description as an idea of a local tinkerer, who cannibalised a drone for its AI modules, adapted or wrote their own software and made it run the Boxwing. Then shared the information how it was done, and the local shipyards started doing it en masse. (I guess the price making sense would then be that of the Boxwing + part of the drone's price)

**Reply by lumbar527** [reply]
According to the fandom, the Boxwing is sold by Southbound Shipyards and Tarazed Corporation.

**Reply by ghost** [reply]
I tried finding the information, about what's sold where (or rather, as a part of which group) for Boxwing - as in, which group(s) it belongs to. That information seems to be in the game's \data\human\sales.txt file, but Boxwing is not there.

Where can it be found for that specific ship?

Edit:
I found this in \data\human\campaign events.txt file:
event "southern carriers 2"
	shipyard "Southbound Basics"
		"Boxwing"
(with indents, they don't seem to be displayed here properly)

I guess that means, that after this specific event happens, Boxwing is added to Southbound Basics category?
If I wanted to add to the plugin's Boxwing to the same category (after the same event), if I place the same data in the plugin (with a different ship name, of course), will it work to add it? (and will it work retroactively, since the event has already happened in my game?)

Edit 2:
I also found this in my save:
shipyard "Southbound Basics"
		Boxwing

I guess, that I would have to edit it manually, and add the new Boxwing version there?

---

Also a question, to make sure, if I get it right:
If, making a plugin, I place the new Boxwing version in ships.txt file (in the plugin's folder of course), or a file with any other name, it should work equally well, assuming it contains all the necessary and properly formatted data?

**Reply by lumbar527** [reply]
> If I wanted to add to the plugin's Boxwing to the same category (after the same event), if I place the same data in the plugin (with a different ship name, of course), will it work to add it? (and will it work retroactively, since the event has already happened in my game?)

I'm not entirely sure what you're asking here but this is what you need to do:
1. Make the ship.
2. Edit the event/make a new one.
3. Either edit your save or create a new one (I would suggest the latter).
I don't think your save will be automatically updated for an event that would have happened in the past but was not implemented yet.

**Reply by ghost** [reply]
Yeah, that's what I meant - since it won't work retroactively, I'll probably need to edit my save.

Also, can planets have the new Shipyard categories added by events?
I mean, let's say I wanted to add the new Boxwing to the western side of the map's FW planets, but not Tarazed, after the same event, which adds the original Boxwing.
(that is, Glaze in Aldhibain, Zug in Zubeneschamali, and Rust in Kraz)
If I created a new Shipyard group, named it "Autoboxwing Group", and placed, in the
event "southern carriers 2", the following:

planet Glaze:
shipyard "Autoboxwing Group"
planet Zug:
shipyard "Autoboxwing Group"
planet Rust:
shipyard "Autoboxwing Group"

(plus appropriate indents)

Would it be added to the three planets, when the event happened?

**Reply by lumbar527** [reply]
I think it would be something like this:
```
planet Glaze
	add shipyard "Autoboxwing Group"
```
But yes.

**Reply by ghost** [reply]
I was looking at the "campaign events.txt" file, which doesn't have the "add" word, but it's only updating the existing shipyards, not adding new ones.
https://github.com/endless-sky/endless-sky/wiki/CreatingEvents seems to confirm what you wrote:

```
planet <name>
  [(add | remove)] shipyard <name>
```

**Reply by ghost** [reply]
I'm trying to actually get around to it... and since I really don't want to mess up the copyright part,
I have a question:
Do all vanilla game's graphic and sound assets belong to Michael Zahniser, and as such, would a copyright file like this be sufficient?
With actual e-mail there, as provided in the example plugin, I removed it from here, in case some bots are out e-mail hunting and don't have it yet. (Hey, the fact I'm paranoid doesn't mean there isn't an invisible demon about to eat my face. ;) )
```
Format: http://www.debian.org/doc/packaging-manuals/copyright-format/1.0/

Files: images/*
Copyright: Michael Zahniser <e-mail>
License: CC-BY-SA-4.0

Files: sounds/*
Copyright: Michael Zahniser <e-mail>
License: CC-BY-SA-4.0
```

As I'd probably use the graphic and sound assets of the original Boxwing, but nothing else, unless I'm missing something at this moment.

**Reply by Hecter94** [reply]
You might want to check out our community [discord](https://discord.gg/ZeuASSx).
You don't have to, of course, but a lot of our most knowledgable plugin authors are quite active on Discord without ever really checking GitHub, so you'll likely get better help there. 

If you do join, check out #plugin-chat and ask any questions you may have there, and you should get prompt replies.

**Reply by ghost** [reply]
Yeah, I probably will.
Creating a ship worked fine, and after I added it to "Basic Ships" group, it was visible on New Boston after starting a new game.

Where I'm stuck, is creating the event to make it available together with a regular Boxwing.
(well, and I also need to write a new description, but that won't be a problem, once I get around to it)

I tried various ways to do it, and so far none of them worked.

This part was supposed to go into final version, if it worked (which I would test on a copy of my regular save, if the test one worked first).

```
shipyard "Autoboxwing Group"
	"Autoboxwing"
	
event "southern carriers 2"
	planet "Glaze"
		shipyard "Autoboxwing Group"
	planet "Zug"
		shipyard "Autoboxwing Group"
	planet "Rust"
		shipyard "Autoboxwing Group"		
```
This was a test attempt, to make it appear on New Boston a day after starting a new game, and it fails to work:
(either with or without "add", I tried both just in case)

```
event "TestBoxwingAdd"
	date 17 11 3013
		shipyard "Basic Ships"
			add "Autoboxwing"
```
or
```
		shipyard "Basic Ships"
			"Autoboxwing"
```

Trying to add the new shipyard group didn't work, either:
```
event "TestBoxwingAdd"
	date 17 11 3013
	planet "New Boston"
		shipyard "Autoboxwing Group"
```
or
```
event "TestBoxwingAdd"
	date 17 11 3013
	planet "New Boston"
		add shipyard "Autoboxwing Group"
```

**Reply by lumbar527** [reply]
I'm not sure what's wrong. Sometimes when I'm making a plugin events/missions refuse to work and I don't know why. I have a full storyline (I don't think it was that good though) that later missions just don't pop up.
